### ISSUE TITLE
MueLu_Test_ETI.hpp does not respect Tpetra instantations

### ISSUE THREAD

**Issue by mhoemmen** [trilinos/Trilinos#1724]
@trilinos/muelu @trilinos/xpetra @trilinos/tpetra 

`muelu/test/unit_tests/MueLu_Test_ETI.hpp` instantiates a test for enabled (LO, GO, Node) combinations.  It does check Tpetra macros for whether Node types are enabled, but it does _not_ correctly check Tpetra macros for whether `Scalar=double` or given (LO, GO) type combinations are enabled.  This causes build errors, e.g., when turning off `GO=int` to speed up a CUDA build.

Also, `muelu/src/Utils/MueLu_ETI_4arg.hpp` instantiates Tpetra for Serial or OpenMP Node types, even if neither of those Node types is enabled.  (It would be nice to speed up CUDA builds by insisting that only the CUDA Node type get instantiated.)

I'm working on a fix now.

**Comment by mhoemmen** [comment]
@brian-kelley This is why I haven't gotten to final tests of your PR #1606 yet ;-) .

**Comment by mhoemmen** [comment]
Argh, OK, my fix didn't fix everything.  I'll just turn on all those Node and GO types that I don't really need to make MueLu happy so that I can actually run tests.

**Comment by cgcgcg** [comment]
There have been no changes to `muelu/test/unit_tests/MueLu_Test_ETI.hpp` since the bug was reported, so I assume the problem is still there?
I can see that `KOKKOS_HAVE_CUDA`, `HAVE_TPETRA_INST_CUDA` and `HAVE_MUELU_INST_DOUBLE_INT_INT` are checked before `double-int-int` is enabled. Is that not sufficient? Can you point me to a configuration that fails?

**Comment by mhoemmen** [comment]
@cgcgcg I haven't tried this for a while so I'm not sure.  The best way to test would be to set `TPETRA_INST_INT_INT:BOOL=OFF` and see if everything builds.  I think the MueLu tutorial may require `GO=int` (I'm not sure why -- could it have something to do with Matlab?), so you have have to disable the tutorial too.

**Comment by brian-kelley** [comment]
I think this issue can be closed. Even though MueLu_Test_ETI.hpp doesn't check ``defined(HAVE_TPETRA_INST_<LO>_<GO>)``, it does check the MueLu Scalar/LO/GO combinations like ``defined(HAVE_MUELU_INST_DOUBLE_INT_INT)``, and those CMake variables are computed in muelu/CMakeLists.txt based on the Tpetra ETI settings. We went through and solved the GO=int vs. GO=long_long issues when we refactored Tpetra to enable only a single GO per build.

For the second point about MueLu_ETI_4arg.hpp, it does instantiate for Serial or OpenMP, but only if MueLu+Epetra is enabled. If only Tpetra is enabled, then the only ETI happens through the macro ``TPETRA_INSTANTIATE_SLGN_NO_ORDINAL_SCALAR`` (the types actually enabled in Tpetra).
